BaseLib/ConfigTreeNew.cpp

```diff
@@ -0,0 +1,225 @@
+    : _tree(&tree), _path(joinPaths(parent._path, root)),
+      _onerror(parent._onerror), _onwarning(parent._onwarning)
+{
+    checkKeyname(root);
+}
+
+ConfigTreeNew::
+ConfigTreeNew(ConfigTreeNew && other)
+	: _tree(other._tree)
+	, _path(other._path)
+	, _visited_params(std::move(other._visited_params))
+	, _onerror(other._onerror)
+	, _onwarning(other._onwarning)
+{
+	other._tree = nullptr;
+}
+
+ConfigTreeNew::~ConfigTreeNew()
+{
+	if (!_tree) return;
+
+	for (auto const& p : *_tree)
+	{
+		markVisitedDecrement(p.first);
+	}
+
+	for (auto const& p : _visited_params)
+	{
+		if (p.second.count > 0) {
+			warning("Key <" + p.first + "> has been read " + std::to_string(p.second.count)
+					+ " time(s) more than it was present in the configuration tree.");
+		} else if (p.second.count < 0) {
+			warning("Key <" + p.first + "> has been read " + std::to_string(-p.second.count)
+					+ " time(s) less than it was present in the configuration tree.");
+		}
+	}
+}
+
+ConfigTreeNew
+ConfigTreeNew::
+getConfSubtree(std::string const& root)
+{
+    if (auto t = getConfSubtreeOptional(root)) {
+        return std::move(*t);
+    } else {
+        error("Key <" + root + "> has not been found.");
+        return ConfigTreeNew(PTree(), *this, ""); // TODO that will crash
+    }
+}
+
+boost::optional<ConfigTreeNew>
+ConfigTreeNew::
+getConfSubtreeOptional(std::string const& root)
+{
+    checkUnique(root);
+    auto subtree = _tree->get_child_optional(root);
+
+    if (subtree) {
+        markVisited(root);
+        boost::optional<ConfigTreeNew> t;
+        t.emplace(*subtree, *this, root);
+        return std::move(t);
+    } else {
+        markVisited(root, true);
+        return boost::optional<ConfigTreeNew>();
+    }
+}
+
+Range<ConfigTreeNew::SubtreeIterator>
+ConfigTreeNew::
+getConfSubtreeList(std::string const& root)
+{
+    checkUnique(root);
+    markVisited(root, true);
+
+    auto p = _tree->equal_range(root);
+
+    return Range<SubtreeIterator>(
+                SubtreeIterator(p.first,  root, *this),
+                SubtreeIterator(p.second, root, *this));
+}
+
+void ConfigTreeNew::ignoreConfParam(const std::string &param)
+{
+    checkUnique(param);
+    // if not found, peek only
+    bool peek_only = _tree->find(param) == _tree->not_found();
+    markVisited(param, peek_only);
+}
+
+void ConfigTreeNew::ignoreConfParamAll(const std::string &param)
+{
+    checkUnique(param);
+    auto& ct = markVisited(param, true);
+
+    auto p = _tree->equal_range(param);
+    for (auto it = p.first; it != p.second; ++it) {
+        ++ct.count;
+    }
+}
+
+
+void ConfigTreeNew::error(const std::string& message)
+{
+	_onerror(_path, message);
+}
+
+void ConfigTreeNew::warning(const std::string& message)
+{
+	_onwarning(_path, message);
+}
+
+
+void ConfigTreeNew::onerror(const std::string& path, const std::string& message)
+{
+    ERR("ConfigTree: At path <%s>: %s", path.c_str(), message.c_str());
+    std::abort();
+}
```

Comment:
I think in this special case there would not be much benifit from it: onerror() is wrapped inside a std::function. Does std::function have special support for noreturn? So far I would prefer to not do the extra work.